File: training-6.0/6-3j.cpp
// https://contest.yandex.ru/contest/66794/problems/J/
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <deque>
using namespace std;

int main() {
	const size_t maxValue = 1000000001;
	ifstream cin("input.txt");
	size_t n, height;
	cin >> n >> height;
	vector<pair<size_t, size_t>> arr(n);
	for (size_t i = 0; i < n; i++)
		cin >> arr[i].first; // H
	for (size_t i = 0; i < n; i++)
		cin >> arr[i].second; // W
	arr.push_back({ 10 * maxValue, 0 });
	sort(arr.begin(), arr.end());
	deque<pair<size_t, size_t>> dq; // index, discomfort
	dq.push_back({ 0,0 });
	size_t l = 0, r = 1, lastHeight = arr[0].first;
	unsigned long long sumWidth = arr[0].second,discomfort = 0, res = maxValue;
	while (l < n) {
		while (r <= n and sumWidth < height) {
			discomfort = arr[r].first - lastHeight;
			while (!dq.empty() and dq.back().second <= discomfort)
				dq.pop_back();
			sumWidth += arr[r].second;
			dq.push_back({ r, discomfort });
			lastHeight = arr[r].first;
			r++;
		}
		auto head = dq.front();
		discomfort = head.second;
		if (discomfort < res) {
			res = discomfort;
			if (res == 0) break;
		}
		sumWidth -= arr[l].second;
		l++;
		if (l >= head.first)
			dq.pop_front();

	}
	cout << res << '\n';
	return 0;
}

File: training-6.0/6-3j.py
# https://contest.yandex.ru/contest/66794/problems/J/
from collections import deque

reader = open('input.txt', 'r')
n, height = map(int, reader.readline().split())
arr = sorted(zip(map(int, reader.readline().split()), map(int, reader.readline().split())))
arr.append((float('inf'), 0))
reader.close()
dq = deque([(0, 0)])
discomfort, res, l, r = 0, float('inf'), 0, 1
last_height, sum_width = arr[0]
while l < n:
    while r <= n and sum_width < height:
        discomfort = arr[r][0] - last_height
        while len(dq) and dq[-1][1] <= discomfort:
            dq.pop()
        sum_width += arr[r][1]
        dq.append((r, discomfort))
        last_height = arr[r][0]
        r += 1
    res = min(res, dq[0][1])
    if res == 0: break
    sum_width -= arr[l][1]
    l += 1
    if l >= dq[0][0]: dq.popleft();
print(res)
